dawn/test/unit-test/dawn/Optimizer/Passes/TestPassSetCaches.cpp

```diff
@@ -17,91 +17,235 @@
+
+  explicit TestPassSetCaches() {
+    std::shared_ptr<SIR> sir = std::make_shared<SIR>(ast::GridType::Cartesian);
+    context_ = std::make_unique<OptimizerContext>(diag_, options_, sir);
+    dawn::UIDGenerator::getInstance()->reset();
+  }
 
-  void runTest(const std::string& filename, int nStencils, int nMultiStages,
+  void runTest(const std::string& filename, int nMultiStages,
                const std::vector<std::vector<std::string>>& fieldNames,
-               const std::vector<std::vector<iir::Cache::CacheType>>& cacheTypes,
-               const std::vector<std::vector<iir::Cache::IOPolicy>>& ioPolicies) {
-    if(nStencils < 1)
-      nStencils = 1;
+               const std::vector<std::vector<Cache::CacheType>>& cacheTypes,
+               const std::vector<std::vector<Cache::IOPolicy>>& ioPolicies) {
+    std::string filepath = filename;
+    if(!TestEnvironment::path_.empty() && filepath.find('/') != 0)
+      filepath = TestEnvironment::path_ + "/" + filepath;
+    auto instantiation = IIRSerializer::deserialize(filepath);
 
-    std::shared_ptr<iir::StencilInstantiation> instantiation =
-        CompilerUtil::load(filename, options_, context_, TestEnvironment::path_);
+    //    ASSERT_TRUE(CompilerUtil::runGroup(PassGroup::Parallel, context_, instantiation));
+    //    ASSERT_TRUE(CompilerUtil::runGroup(PassGroup::ReorderStages, context_, instantiation));
 
-    ASSERT_TRUE(CompilerUtil::runGroup(PassGroup::Parallel, context_, instantiation));
-    ASSERT_TRUE(CompilerUtil::runGroup(PassGroup::ReorderStages, context_, instantiation));
+    // Run stage splitter pass
+    PassStageSplitter stageSplitPass(*context_);
+    EXPECT_TRUE(stageSplitPass.run(instantiation));
 
     // Expect pass to succeed...
-    ASSERT_TRUE(CompilerUtil::runPass<dawn::PassSetCaches>(context_, instantiation));
+    // ASSERT_TRUE(CompilerUtil::runPass<dawn::PassSetCaches>(context_, instantiation));
+    PassSetCaches setCachesPass(*context_);
+    EXPECT_TRUE(setCachesPass.run(instantiation));
 
     auto& stencils = instantiation->getStencils();
-    ASSERT_EQ(stencils.size(), nStencils);
-
-    for(int i = 0; i < nStencils; ++i) {
-      auto& multiStages = stencils[i]->getChildren();
-      ASSERT_EQ(multiStages.size(), nMultiStages);
-
-      int j = 0;
-      for(const auto& multiStage : multiStages) {
-        unsigned nFields = fieldNames[j].size();
-        for(int k = 0; k < nFields; ++k) {
-          int accessID = stencils[i]->getMetadata().getAccessIDFromName(fieldNames[j][k]);
-          ASSERT_TRUE(multiStage->isCached(accessID));
-          ASSERT_TRUE(multiStage->getCache(accessID).getType() == cacheTypes[j][k]);
-          ASSERT_TRUE(multiStage->getCache(accessID).getIOPolicy() == ioPolicies[j][k]);
-        }
-        j += 1;
+    ASSERT_GT(stencils.size(), 0);
+
+    auto& stencil = stencils[0];
+    auto& multiStages = stencil->getChildren();
+    ASSERT_EQ(multiStages.size(), nMultiStages);
+
+    int i = 0;
```

Comment:
Could we use `enumerate` here? https://github.com/MeteoSwiss-APN/dawn/blob/master/dawn/src/dawn/Support/External/easy_iterator.hpp#L395

Reply:
Done!